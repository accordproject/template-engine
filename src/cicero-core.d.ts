declare module '@accordproject/cicero-core';