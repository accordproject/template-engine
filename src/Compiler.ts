/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import traverse from 'traverse';
import { copySync } from 'fs-extra';

import { TemplateMarkModel, CiceroMarkModel, CommonMarkModel, ConcertoMetaModel } from '@accordproject/markdown-common';

import { ClassDeclaration, Introspector, ModelManager, ModelUtil, Resource } from '@accordproject/concerto-core';
import { CodeGen } from '@accordproject/concerto-tools';
import { FileWriter } from '@accordproject/concerto-util';
import { getCompiler } from './compilers/NodeCompilers';
import { RUNTIME_DIR, writeEpilog, writeImports, writeProlog } from './compilers/Common';

export type ProcessingFunction = (fw:FileWriter, level:number, resource:any) => void;

export type CompilationOptions = {
    skipGenerator?: boolean;
    skipCopyRuntime?: boolean;
}

/**
 * Converts TemplateMark DOM to a Typescript function, that, when run with JSON data
 * that conforms to the template model will produce AgreementMark JSON output.
 */
export class Compiler {
    modelManager: ModelManager;
    templateClass: ClassDeclaration;

    constructor(modelManager: ModelManager) {
        this.modelManager = modelManager;
        const introspector = new Introspector(this.modelManager);
        const templateModels = introspector.getClassDeclarations().filter((item) => {
            return !item.isAbstract() && item.getDecorator('template');
        });
        if (templateModels.length > 1) {
            throw new Error('Found multiple concepts with @template decorator. The model for the template must contain a single concept with the @template decorator.');
        } else if (templateModels.length === 0) {
            throw new Error('Failed to find a concept with the @template decorator. The model for the template must contain a single concept with the @template decoratpr.');
        } else {
            this.templateClass = templateModels[0];
        }
    }

    writeFunction(fw: FileWriter, functionName: string, returnType:string, code: string) {
        fw.writeLine(0, `export function ${functionName}(data:TemplateModel.I${this.templateClass.getName()}, library:any) : ${returnType} {`);
        this.templateClass.getProperties().forEach((p: any) => {
            fw.writeLine(1, `const ${p.getName()} = data.${p.getName()};`);
        });
        fw.writeLine(1, 'const now = dayjs();');
        fw.writeLine(1, code.trim());
        fw.writeLine(0, '}');
        fw.writeLine(0, '');
    }

    /**
     * Get a model manager with templatemark model
     * @param {string} outputDir the output directory
     * @returns {ModelManager} the model manager
     */
    getTemplateMarkModelManager(): ModelManager {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel.MODEL, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel.MODEL, 'commonmark.cto');
        modelManager.addCTOModel(TemplateMarkModel.MODEL, 'templatemark.cto');
        return modelManager;
    }

    /**
     * Get a model manager with templatemark model
     * @param {string} outputDir the output directory
     * @returns {ModelManager} the model manager
     */
    getCiceroMarkModelManager(): ModelManager {
        const modelManager = new ModelManager({ strict: true });
        modelManager.addCTOModel(ConcertoMetaModel.MODEL, 'concertometamodel.cto');
        modelManager.addCTOModel(CommonMarkModel.MODEL, 'commonmark.cto');
        modelManager.addCTOModel(CiceroMarkModel.MODEL, 'ciceromark.cto');
        return modelManager;
    }

    generateTypeScript(modelManager: ModelManager, outputDir: string) {
        // generate the TS code for the template model
        const tsVisitor = new CodeGen.TypescriptVisitor();
        const parameters = {
            fileWriter: new FileWriter(outputDir)
        };
        modelManager.accept(tsVisitor, parameters);
    }

    static doIt(fw:FileWriter, level:number, templateMark: any) {
        fw.writeLine(level, `// start processing ${ModelUtil.getShortName(templateMark.$class)}`);
        traverse(templateMark).forEach(function (context: any) {
            if (typeof context === 'object' && context.$class && typeof context.$class === 'string') {
                const nodeClass = context.$class as string;

                const compiler = getCompiler(nodeClass);
                const sourceLevel = level+1;

                this.before( () => {
                    compiler.enter(fw, sourceLevel, context);
                    compiler.generate(fw, sourceLevel, context, Compiler.doIt);
                });

                this.after( () => {
                    compiler.exit(fw, sourceLevel, context);
                });
            }
        });
        fw.writeLine(level, `// end processing ${ModelUtil.getShortName(templateMark.$class)}`);
    }
    /**
     * Converts a TemplateMark document to typescript code that is parameterised
     * by the 'data' argument.
     *
     * The strategy is to traverse the template mark json, converting each node into
     * typescript statements that add an equivalent ciceromark node. Variables will
     * pull values from 'data'.
     *
     * Conditionals and formulae will call the functions produced by
     * compileUserCode.
     *
     * Note that code generation is itself recursive, for example, when
     * a list definition node is encountered the list must be iterated over
     * based on a variable value coming from data, and code must be
     * generated to produce each list item.
     *
     * '$curNode' is the current ciceromark node that is being populated.
     * Typescript lexical scoping (blocks) is used to introduce local
     * variables.
     *
     * All internal variables are prefixed with '$' to ensure (?)
     * that they cannot collide with end user variable names.
     *
     * @param {IDocument} templateMark the TemplateMark Concerto resource
     * @param {string} outputDir the output directory for code
     */
    compileGenerator(templateMark: Resource, outputDir: string) : void {
        const fw = new FileWriter(outputDir);

        fw.openFile('generator.ts');
        fw.writeLine(0, '// generated code, do not modify');
        fw.writeLine(0, '');

        fw.writeLine(0, '// IMPORTS');
        writeImports(fw, 0, this.templateClass);
        fw.writeLine(0, '');

        fw.writeLine(0, '// GENERATOR');
        writeProlog(fw, 0, this.templateClass);

        Compiler.doIt(fw, 1, templateMark);

        writeEpilog(fw,0);
        fw.closeFile();
    }

    copyRuntime(outputDir: string) {
        copySync('./src/drafting', `${outputDir}/${RUNTIME_DIR}/drafting`);
        copySync('./src/Runtime.ts', `${outputDir}/${RUNTIME_DIR}/Runtime.ts`);
    }

    compile(templateMark: Resource, outputDir: string, options?:CompilationOptions) {
        // generate the model
        this.generateTypeScript(this.modelManager, `${outputDir}`);
        this.generateTypeScript(this.getTemplateMarkModelManager(), `${outputDir}`);
        this.generateTypeScript(this.getCiceroMarkModelManager(), `${outputDir}`);

        // generate names for all the nodes containing user code
        const namedTemplateMark = this.nameUserCode(templateMark);

        // compile the user code
        this.compileUserCode(namedTemplateMark, outputDir);

        // compile the generator
        if(!options?.skipCopyRuntime) {
            this.copyRuntime(outputDir);
        }

        // compile the generator
        if(!options?.skipGenerator) {
            this.compileGenerator(namedTemplateMark, outputDir);
        }
    }

    nameUserCode(templateMark: Resource) {
        const templateMarkDom = templateMark.toJSON();
        return traverse(templateMarkDom).map(function (x) {
            if (x && ((x.$class === `${TemplateMarkModel.NAMESPACE}.ConditionalDefinition` && x.condition) ||
                (`${TemplateMarkModel.NAMESPACE}.ClauseDefinition` && x.condition))) {
                x.functionName = `condition_${this.path.join('_')}`;
            }
            this.update(x);
        });
    }

    compileUserCode(templateMark: Resource, outputDir: string) {
        const fw = new FileWriter(outputDir);

        fw.openFile('usercode.ts');
        fw.writeLine(0, '// generated code, do not modify');
        fw.writeLine(0, '');

        fw.writeLine(0, '// IMPORTS');
        fw.writeLine(0, `import * as TemplateModel from './${this.templateClass.getNamespace()}';`);
        fw.writeLine(0, 'import dayjs from \'dayjs\';');
        fw.writeLine(0, 'import jp from \'jsonpath\';');
        fw.writeLine(0, '');

        const templateMarkDom = templateMark;

        const functionNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.FormulaDefinition`) {
                acc.push({
                    name: x.name,
                    code: x.code,
                });
            }
            return acc;
        }, []);

        const conditionalNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.ConditionalDefinition` && x.condition) {
                acc.push({
                    name: x.functionName,
                    code: x.condition,
                });
            }
            return acc;
        }, []);

        const clauseNodes: [] = traverse(templateMarkDom).reduce(function (acc, x) {
            if (x && x.$class === `${TemplateMarkModel.NAMESPACE}.ClauseDefinition` && x.condition) {
                acc.push({
                    name: x.functionName,
                    code: x.condition,
                });
            }
            return acc;
        }, []);

        fw.writeLine(0, '// FORMULAE');
        functionNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'any', fun.code.contents);
        });
        fw.writeLine(0, '');

        fw.writeLine(0, '// CONDITIONALS');
        conditionalNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'boolean', `return ${fun.code.contents}`);
        });
        fw.writeLine(0, '');

        fw.writeLine(0, '// CLAUSES');
        clauseNodes.forEach((fun: any) => {
            this.writeFunction(fw, `${fun.name}`, 'boolean', `return ${fun.code.contents}`);
        });
        fw.writeLine(0, '');

        fw.closeFile();
    }
}