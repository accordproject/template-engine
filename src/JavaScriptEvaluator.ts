/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import child_process from 'child_process';
import jp from 'jsonpath';
import dayjs from 'dayjs';
import { EventEmitter } from 'node:events';
import path from 'path';

export type EvalOptions = {
    timeout: number // timeout in milliseconds for this eval request
}

export type EvalRequest = {
    code: string, // JS code to eval
    argumentNames: string[] // names of function args
    arguments: any[] // function arg values, these have to be serializable to JSON!
}

type WorkItem = {
    pid?: number; // process if assigned to the workitem
    startTime: number; // timestamp when workitem created
    expireTime: number; // timestamp for when work must be completed
    request: EvalRequest; // payload to process
    resolve: (result: any) => void; // promise callback for success
    reject: (result: any) => void; // promise callback for failure
}

export type EvalResponse = {
    result: any // success if not null
    timeout?: boolean // if true the promise is rejected due to timeout
    starvation?: boolean // if true the promise is rejected due to lack of child process
    message?: string; // if promise rejected due to a caught exception this will be set
    elapsed?: number; // the elapsed time in ms to process the work item
    maxQueueDepthExceeded?: boolean // if true the promise is rejected because the queue is full
}

export type JavaScriptEvaluatorOptions = {
    waitInterval: number; // time to wait in ms for a free worker
    maxWorkers: number; // the max number of worker processes
    maxQueueDepth: number; // the max queue depth for the waiting queue
}

type ChildProcess = {
    pid?: number
} & EventEmitter;

async function sleep(msec: number) {
    return new Promise(resolve => setTimeout(resolve, msec));
}

/**
 * This class implements two JS function evaluation strategies:
 * 1. evalDangerously which creates a dynamic function and run it in-process
 * This should only be used with trusted code, or within a sandbox (e.g. the browser)
 * 2. evalChildProcess which spins up a child Node process to eval the function
 * The maximum number of child processes is capped via JavaScriptEvaluatorOptions
 * as well as the maximum queue depth for the queue used to wait for a free worker
 * child process. Not that to prevent cross-request contamination
 * child processes are NOT pooled and are never reused.
 */
export class JavaScriptEvaluator {
    options: JavaScriptEvaluatorOptions;
    workers: Array<ChildProcess>; // child processes
    queue: Array<WorkItem>; // queue of work to do

    constructor(options: JavaScriptEvaluatorOptions = {waitInterval: 50, maxWorkers: 8, maxQueueDepth: 1000}) {
        this.options = options;
        this.workers = [];
        this.queue = [];
    }

    /**
     * Evaluates a JS function in process.
     * @param {EvalRequest} request - the eval request
     * @returns {Promise} a promise to the result
     */
    async evalDangerously(request: EvalRequest): Promise<EvalResponse> {
        return new Promise((resolve, reject) => {
            try {
                const start = new Date().getTime();
                const fun = new Function(...['dayjs', 'jp', ...request.argumentNames], request.code);
                const result = fun(...[dayjs, jp, ...request.arguments]);
                const end = new Date().getTime();
                resolve({ result, elapsed: end - start });
            }
            catch (err: any) {
                reject({
                    message: err.message
                });
            }
        });
    }
    /**
     * Evaluates a JS function using a node child process
     * @param {EvalRequest} request the eval request
     * @param {EvalOptions} options the options for the request
     * @returns {Promise<EvalResponse>} the async result
     */
    async evalChildProcess(request: EvalRequest, options: EvalOptions = { timeout: 10000 }): Promise<EvalResponse> {
        // console.log('Workers count: ' + this.workers.length);
        return new Promise((resolve, reject) => {
            const now = new Date().getTime();
            const workItem: WorkItem = {
                startTime: now,
                expireTime: now + options.timeout,
                request,
                resolve,
                reject
            };
            if(this.queue.length >= this.options.maxQueueDepth) {
                reject({ maxQueueDepthExceeded: true, elapsed: 0 });
            }
            this.queue.push(workItem);
            this.processQueue(options);
        });
    }
    private processQueue(options: EvalOptions) {
        const now = new Date().getTime();
        const notExpired = this.queue.filter(w => (now < w.expireTime));
        const expired = this.queue.filter(w => (now >= w.expireTime));
        this.queue = notExpired;
        expired.forEach(w => {
            w.reject({ timeout: true, starvation: true, elapsed: now - w.startTime });
        });
        if (this.workers.length < this.options.maxWorkers) {
            // we have a free worker
            const next = this.queue.shift();
            if (next) {
                this.doWork(next, options)
                    .then(result => next.resolve(result))
                    .catch(error => next.reject(error));
            }
        }
        else {
            // no free worker, so sleep and then try again
            sleep(this.options.waitInterval)
                .then(() => {
                    this.processQueue(options);
                });
        }
    }
    private getWorkerPath() : string {
        try {
            const thisPath = require.resolve('@accordproject/template-engine');
            return path.join(thisPath,'..','worker.js');
        } catch(err) { // eslint-disable-line @typescript-eslint/no-unused-vars
            // if we cannot load the module it likely means we are running
            // a unit test inside the module...
            return path.join(__dirname, '..', 'dist', 'worker.js');
        }
    }
    private doWork(work: WorkItem, options: EvalOptions): Promise<EvalResponse> {
        return new Promise((resolve, reject) => {
            const start = new Date().getTime();
            // check for browser
            if (!child_process.fork) {
                reject({ message: 'Cannot use evalChildProcess because child_process.fork is not defined.' });
            }
            // on timeout will send SIGTERM
            const workerPath = this.getWorkerPath();
            // console.debug(`Worker path: ${workerPath}`);
            const worker = child_process.fork(workerPath, { timeout: options.timeout, env: {} });
            if (!worker.pid) {
                throw new Error('Failed to fork child process');
            }
            this.workers.push(worker);
            work.pid = worker.pid;
            let result: any;
            worker.on('error', (err: any) => {
                this.workers = this.workers.filter((w: ChildProcess) => w.pid !== worker.pid);
                const end = new Date().getTime();
                reject({ message: err.message, elapsed: end - start });
            });
            worker.on('message', (msg: any) => {
                result = msg;
            });
            worker.on('exit', (code: any) => {
                if (code === null) {
                    this.workers = this.workers.filter((w: ChildProcess) => w.pid !== worker.pid);
                    const end = new Date().getTime();
                    reject({ timeout: true, elapsed: end - start });
                }
                else if (code === 0 && result) {
                    // success!
                    this.workers = this.workers.filter((w: ChildProcess) => w.pid !== worker.pid);
                    const end = new Date().getTime();
                    resolve({ ...result, elapsed: end - start });
                } else {
                    // null result or non-zero code from worker means an error
                    // result will be undefined if the user code calls process.exit()
                    this.workers = this.workers.filter((w: ChildProcess) => w.pid !== worker.pid);
                    const end = new Date().getTime();
                    reject({ code, result, elapsed: end - start });
                }
            });
            // send the request to the child process
            worker.send(work.request);
        });
    }
}
