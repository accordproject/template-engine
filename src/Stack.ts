/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
interface IStack<T> {
    push(item: T): void;
    pop(): T | undefined;
    peek(): T | undefined;
    size(): number;
  }

class Stack<T> implements IStack<T> {
    private storage: T[] = [];

    constructor(private capacity: number = Infinity) {}

    push(item: T): void {
        if (this.size() === this.capacity) {
            throw Error('Stack has reached max capacity, you cannot add more items');
        }
        this.storage.push(item);
    }

    pop(): T | undefined {
        return this.storage.pop();
    }

    peek(): T | undefined {
        return this.storage[this.size() - 1];
    }

    size(): number {
        return this.storage.length;
    }
}