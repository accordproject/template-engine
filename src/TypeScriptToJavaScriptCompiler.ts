/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { createDefaultMapFromNodeModules, createDefaultMapFromCDN } from '@typescript/vfs';
import { twoslasher, TwoSlashOptions, TwoSlashReturn } from '@typescript/twoslash';
import { ModelManager } from '@accordproject/concerto-core';
import { TypeScriptCompilationContext } from './TypeScriptCompilationContext';
import { DAYJS_BASE64, JSONPATH_BASE64 } from './runtime/declarations';
import * as lzstring from 'lz-string';

/**
 * Compiles user Typescript code to JavaScript. This uses the '@typescript/twoslash'
 * project which is maintained by the Typescript team and powers their web playground.
 * It uses the TypeScriptCompilationContext class to construct a payload for twoslash
 * that is composed of multiple TS files to compile, along with their 3rd-party module
 * dependencies.
 *
 * Note that the 'typescript' module is either dynamically loaded from node_modules (Node.js)
 * or from the CDN (browser). This module is used by twoslash.
 *
 * The 'updateRuntimeDependencies' script it used to package type declarations for 3rd-party
 * modules that we need to expose to user TS code: dayjs and jsonpath, these also need to be
 * added to the twoslash compilation context.
 */
export class TypeScriptToJavaScriptCompiler {
    context: string;
    fsMap: Map<string,string>|undefined;
    ts: any;

    constructor(modelManager: ModelManager, templateConceptFqn?: string) {
        this.context = new TypeScriptCompilationContext(modelManager, templateConceptFqn).getCompilationContext();
    }

    async initialize() {
        if(typeof window === 'undefined') {
            this.ts = await import ('typescript');
            this.fsMap = createDefaultMapFromNodeModules({
                target: this.ts.ScriptTarget.ES2020,
            });
        }
        else {
            this.ts = (await import('https://cdn.jsdelivr.net/npm/typescript@4.8.4/+esm')).default;
            if(!this.ts) {
                throw new Error('Failed to dynamically load typescript');
            }
            this.fsMap = await createDefaultMapFromCDN({ target: this.ts.ScriptTarget.ES2020 }, this.ts.version, false, this.ts);
        }
        this.fsMap.set('/node_modules/@types/dayjs/index.d.ts', Buffer.from(DAYJS_BASE64, 'base64').toString());
        this.fsMap.set('/node_modules/@types/jsonpath/index.d.ts', Buffer.from(JSONPATH_BASE64, 'base64').toString());
    }

    compile(typescript: string): TwoSlashReturn {
        if(!this.fsMap) {
            throw new Error('initialize must be awaited before compile is called.');
        }
        const twoSlashCode =`
${this.context}
${typescript}
`;

        const options: TwoSlashOptions = {
            fsMap: this.fsMap,
            tsModule: this.ts,
            lzstringModule:lzstring,
            defaultOptions: {
                showEmit: true,
                noErrorValidation: true,
                showEmittedFile: 'code.js'
            }
        };
        const result = twoslasher(twoSlashCode, 'ts', options);
        return result;
    }
}