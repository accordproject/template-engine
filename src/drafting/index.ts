/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import booleanDrafter from './Boolean';
import dateTimeDrafter from './DateTime';
import doubleDrafter from './Double';
import integerDrafter from './Integer';
import longDrafter from './Long';
import monetaryAmountDrafter from './MonetaryAmount';
import stringDrafter from './String';

export function getDrafter(typeName: string) : ((value:any, format:string) => string)|null  {
    switch(typeName) {
    case 'Boolean': return booleanDrafter;
    case 'DateTime': return dateTimeDrafter;
    case 'Double': return doubleDrafter;
    case 'Integer': return integerDrafter;
    case 'Long': return longDrafter;
    case 'org.accordproject.money@0.3.0.MonetaryAmount': return monetaryAmountDrafter;
    case 'String': return stringDrafter;
    default: return null;
    }
}