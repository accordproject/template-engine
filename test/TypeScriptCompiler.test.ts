import { ModelManager } from '@accordproject/concerto-core';
import { readFileSync } from 'fs-extra';
import { TypeScriptToJavaScriptCompiler } from '../src/TypeScriptToJavaScriptCompiler';

describe('typescript compiler', () => {
    test('should compile typescript to javascript', async () => {

        const code = `
export function condition_nodes_0_nodes_13_nodes_4_nodes_0(data:TemplateModel.ITemplateData, library:any) : boolean {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    return preferences.favoriteColors !== undefined && preferences.favoriteColors.includes('PINK')
}`;
        const modelManager = new ModelManager();
        modelManager.addCTOModel( readFileSync('./test/templates/full/model.cto', 'utf-8'), 'model.cto');
        const compiler = new TypeScriptToJavaScriptCompiler(modelManager);
        const results = compiler.compile(code);
        expect(results).toMatchSnapshot();
    });
});