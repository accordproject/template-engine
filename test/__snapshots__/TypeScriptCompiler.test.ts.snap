// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript compiler should compile typescript to javascript 1`] = `
Object {
  "code": "export function condition_nodes_0_nodes_13_nodes_4_nodes_0(data, library) {
    const firstName = data.firstName;
    const lastName = data.lastName;
    const middleNames = data.middleNames;
    const active = data.active;
    const lastVisit = data.lastVisit;
    const address = data.address;
    const orders = data.orders;
    const loyaltyStatus = data.loyaltyStatus;
    const preferences = data.preferences;
    return preferences.favoriteColors !== undefined && preferences.favoriteColors.includes('PINK');
}
",
  "errors": Array [],
  "extension": "js",
  "highlights": Array [],
  "playgroundURL": "https://www.typescriptlang.org/play/#code/FDD0oAgAQMwSwDYFMB2BDAtkgXBAxgPYp5IBOALgVAIwB0ADA7eQM5gBUESLCcK5AWgAmcFmgBGyaOQCeAB255ScOYO69+oFAQFIMqmQL7kyMNCQjtQYSAHFUZNCaH4CQpBBgFSEdFhZy5jiuxGSUNEz0IOAQcPrerNGQAOpopCh8AOa4AEJIAO5pHgQw+HCkeACuCGkQ7goo7sRw3BD5ABaoEBhucDAyWRDknSwecXIJbOMJQ-IeAN7AAJAAkgCCQkKk3CwANMsrAPKk7qT7qwAyBDJoCLIAyuROlXsHAArbMGSoJK9L9o0yOcVgAVPRyGomAAiTjQwAAvp5SAQMBAAOS0UAmFjkCKMehogDcSVi-FMQTYSAAHhMKKSTKQzBYVgBhIgkVQQRYQHkAEjwNRYLFwOOUKEyxPhIGptPIswUEDZoVUAFUMkQIABeCDrTbbIUQAA+EGARxOZCNJpWVxudxkj2eLEtpo+SC+21CTuNprB+khSBhT2JwBlM2M5OZayFSDl1JMjSdrPZSE53L5cCa5D6LVIIvIYolCOlNLDZMZQR1bzSWbwKjQ-C4VPjQkTSo5crTEF5GdQWfgZDzBclxdl9IjHlBpHrYjwWY1cdQLZ1bZTHeAPK7Wf8T30uEDSGHIZLdPD5eZAFEAG69xvN1vJ1Prvlb7g7uR7pwHosgGzQeDIPxgkIUIKAIZg2FATh1GMYRRAkKQoFkBQWCUFQ1B4YwtB0cFZCMMsmQ8KxfwBRxnFcdxPG8XxMG4QISFwYCSFAklpgoNhf1PAjKWPOVOIrJNlTXDd+UFYUIFFLJD1DE98P4qNRljJtF3vQSuSfLse34bMB3E-NJO-aTeNk5kqwoOBa0CBsFwTZcHyE9NM203NdKHAyeLHM8JxBKcUBnOcUFvZTbNUxYpSPUc+Iva8rKUmyBPbNSwpJWBEFQGjcGxXE6HxcCOC4DD+FgsRJA8RC5hQ5RVF0AryCw3R9FwyLCOsGISKnMjCAorwfEAgIggy188QYFj4jY4BWI7eLV0RGBkVRDFQEYsIqGyhgiRY4zuHC0sGQInUNi2HZArildHw3UUkBjQd9I3WtZGu8ViQ3AAvFQHsLMLDI8vazVOY6lyms6eRYABrSp3qenlMAISp+FwFBKgwcQyCk9ymp1G1bgeJ5yBef6VISztkGvBAIbciLNsrT5vk9fHgsJ9SzEvbw4BMNkEG8FgAH53oAbQAXVR0dUERiAAT+zsAFk1guc8tXRaXZbRfYNwAMXPRW5e1NF1c15X1IAOUOEEAH0oRWe4WQuQ57nPKF5bRI3TfNy3rdtqF9c+tHKdBcF-UDNA6cB+zPHKHEDfSlybp5QVyAjrAyY3DAMyEZB4+4PnBfUzJFzIHmxdz0hIYgcws2vXBxAIAhkHrYvY4ANVEVmPxMYu0D1HZcF1Q6hWL7xTjE36yAFuvrix+0cZefPrTHu0HVxlhi7kamPV+LvXXdH5uEPZL-zShPyKQXKJssEunV9CFP0ltwkAQJEUXRTFMqGgliRPoQ0BkAArJ1ZoftEP7fxYOtE+X85D33mj-IggRhjrWAEhDwbUnBwCIIcVQKDfLy07L4Ag+QeYSXFCrDcHM8C3CQPgvS4pvzbTpDAWGs4MEhBEP5E22h3AsBNvQVhN8OHUAAMzcPYSbAALII7gnCAAUH8njYAvv6a+7gEC0F9n6T8AddgQF4OIKcpAZDYHrDIAAlBACuVca4BWwcBHEodSDhxovLaRaBaDwFsXHGixcQjWNjunBxsJaDePcepTxcpk6bDTjRJ02pHG0FCanJA6dF5BKsXKUucBry+KeLQVJ14PHJM0WgHEjcWCswyU4huTdyC5KINY9uPdIl1D8bU-UiSNzBIgP3Mg9TokdNsVU3ycoOa2mxo6Up-jZ7DIXn06xy83Q01+KMmZm9PQeO2LjdIEBFlzO4M4tAzNlBs2rpzCAABCTU2pYbuHgCgJALgABktyNkry3iwHZezWZIHZpzWgfABSVHYRItEbwVgGwANJokMUWIAA",
  "queries": Array [],
  "staticQuickInfos": Array [],
  "tags": Array [],
}
`;
